base/formatio.jl

```diff
@@ -0,0 +1,337 @@
+    :default       => "\033[39m",
+    # formats
+    :bold          => "\033[1m",
+    :underline     => "\033[4m",
+    :blink         => "\033[5m",
+    :reverse       => "\033[7m",
+    :hidden        => "\033[8m",
+)
+
+for i in 0:255
+    text_colors[i] = "\033[38;5;$(i)m"
+end
+
+const disable_text_style = AnyDict(
+    :bold      => "\033[22m",
+    :underline => "\033[24m",
+    :blink     => "\033[25m",
+    :reverse   => "\033[27m",
+    :hidden    => "\033[28m",
+)
+
+# Create a docstring with an automatically generated list
+# of colors.
+let possible_formatting_symbols = [:normal, :bold, :default]
+    available_text_colors = collect(Iterators.filter(x -> !isa(x, Integer), keys(text_colors)))
+    available_text_colors = cat(
+        sort!(intersect(available_text_colors, possible_formatting_symbols), rev=true),
+        sort!(setdiff(  available_text_colors, possible_formatting_symbols));
+        dims=1)
+global const available_text_colors_docstring =
+    string(join([string("`:", key, "`")
+                 for key in available_text_colors],
+                ",\n", ", or \n"))
+end
+
+"""Dictionary of color codes for the terminal.
+
+Available colors are: $available_text_colors_docstring as well as the integers 0 to 255 inclusive.
+
+The color `:default` will print text in the default color while the color `:normal`
+will print text with all text properties (like boldness) reset.
+Printing with the color `:nothing` will print the string without modifications.
+"""
+text_colors
+
+"""
+    printstyled([io], xs...; bold::Bool=false, color::Union{Symbol,Int}=:normal)
+
+Print `xs` in a color specified as a symbol or integer, optionally in bold.
+
+`color` may take any of the values $available_text_colors_docstring
+or an integer between 0 and 255 inclusive. Note that not all terminals support 256 colors.
+If the keyword `bold` is given as `true`, the result will be printed in bold.
+"""
+printstyled(io::IO, msg...; bold::Bool=false, color::Union{Int,Symbol}=:normal) =
+    with_output_color(print, color, io, msg...; bold=bold)
+printstyled(msg...; bold::Bool=false, color::Union{Int,Symbol}=:normal) =
+    printstyled(stdout, msg...; bold=bold, color=color)
+
+
+"""
+    IOFormatBuffer
+
+`IOFormatBuffer` provides a mechanism for recording annotation regions of an IOBuffer.
+These regions can be used to colorize, or otherwise structure or describe, each byte in the output.
+"""
+struct IOFormatBuffer <: AbstractPipe
+    buf::IOBuffer
+    annotation::Vector{Any}
+    starts::Vector{Int}
+    ends::Vector{Int}
+    function IOFormatBuffer()
+        buf = IOBuffer(read=false, write=true, append=true)
+        buf.seekable = false
+        return new(buf, Symbol[], Int[], Int[])
+    end
+end
+
+pipe_writer(io::IOFormatBuffer) = io.buf
+pipe_reader(io::IOFormatBuffer) = error("IOFormatBuffer not readable")
+
+function with_output_color(f::Function, color::Union{Int, Symbol}, io::IO, args...; bold::Bool = false)
```

Comment:
Perhaps a good opportunity to rename `with_output_color` to something less ugly?

Reply:
I had some recollection that we'd already renamed it to `styled` or something like that?

Reply:
We changed `print_with_color` to `printstyled`, but left `with_output_color` unchanged, I forget why.

Reply:
Maybe `withstyle(f, io, color=..., bold=...)`?

Reply:
Or perhaps just `styled(f, io, color=, bold=)`?

Reply:
`bold` being some arbitrary special kwarg should probably be deprecated

Reply:
Why's that? In `printstyled` we have `color` and `bold` keyword arguments and that seems to work quite nicely.